packages/flutter/lib/src/material/text_selection.dart

```diff
@@ -158,11 +160,14 @@ class _MaterialTextSelectionControls extends TextSelectionControls {
     final double x = (endTextSelectionPoint == null)
         ? startTextSelectionPoint.point.dx
         : (startTextSelectionPoint.point.dx + endTextSelectionPoint.point.dx) / 2.0;
-    final double availableHeight
-        = globalEditableRegion.top - MediaQuery.of(context).padding.top - _kToolbarScreenPadding;
-    final double y = (availableHeight < _kToolbarHeight)
-        ? startTextSelectionPoint.point.dy + globalEditableRegion.height + _kToolbarHeight + _kToolbarScreenPadding
-        : startTextSelectionPoint.point.dy - globalEditableRegion.height;
+    final double toolbarHeightNeeded = MediaQuery.of(context).padding.top
+      + _kToolbarScreenPadding
+      + _kToolbarHeight;
+    final double availableHeight = globalEditableRegion.top + endpoints.first.point.dy - textLineHeight;
+    final bool fitsAbove = toolbarHeightNeeded <= availableHeight;
+    final double y = fitsAbove
+        ? startTextSelectionPoint.point.dy - textLineHeight
+        : startTextSelectionPoint.point.dy + _kToolbarHeight + _kToolbarContentDistanceBelow;
     final Offset preciseMidpoint = Offset(x, y);
```

Comment:
Not really related to the issue, but I think `x` is already calculated in `TextSelectionOverlay._buildToolbar`. Instead of doing it here again maybe just use `position.dx`?

Reply:
Yeah seems to be calculating the same exact value to me.  I'll change it.